core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```
@@ -327,7 +327,7 @@ public long currentRows() {
     }
 
     private void openCurrent() {
-      if (partitionKey == null) {
+      if (spec.isUnpartitioned() || partitionKey == null) {
```

Just checking, but I think this only matters for V1 tables right? Shouldn't the key be null for everything else?

---

I have written own implementation of `PartitionedFanoutWriter<T>` when I try to pass the null `partitionKey` from `override fun partition(row: Record)` it fail with NPE at https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/io/PartitionedFanoutWriter.java#L58.
It means we can't pass null `partitionKey`. There's needs to be first class way of knowing whether spec is partitioned or un-partitioned.

---

Why would you use partitioned fan out writer with unpartitioned data?

---

Those would both be new behaviors, either breaking on unpartitioned or changing the layout .

I'm not sure either is really justified. If I had to pick I would just throw an exception on unpartitioned. Since that would be a breaking change, I would probably just do nothing. Users with unpartitioned data to can still use the u partitioned writer if the directory layout is important for them. I said this above, but we shouldn’t need a custom implementation of this class to test whether or not an internal behavior works